handshake.md

```diff
@@ -0,0 +1,583 @@
+- Let `EncryptWithAd()` and `DecryptWithAd()` be the Noise functions of the
+  same names.
+- Let `WriteBytes(bytes)` be a hypothetical function that writes `bytes` bytes
+  over the network to the other host.
+- Let `bytes = ReadBytes(len)` be a hypothetical function that reads `len`
+  bytes over the network from the other host, and returns those bytes as
+  `bytes`.
+ - Let `bytes.slice(start, length)` be a hypothetical function on a byte
+   sequence that returns a slice of a sequence of bytes, starting at position
+   `start`, and including the next `length` bytes.
+ - Let `result = bytes.concat(bytes2)` be a hypothetical function on a byte
+   sequence that concatenates the byte sequence `bytes2` onto the existing byte
+   sequence `bytes`, producing the new byte sequence `result`.
+ - Let `bytes.length` be a hypothetical property of a byte sequence that
+   returns the length of the byte sequence `bytes`, in bytes.
+- Let `min(a, b)` be a hypothetical function that returns the smaller number of
+  `a` and `b`.
+
+### 5.2 Fragmentation
+The maximum Noise message length is 65519 bytes, so any input `plaintext`
+exceeding this length MUST be fragmented into segments, as specified below, in
+order to facilitate encrypted transport.
+
+At a high level, this is done by the following steps:
+
+1. Write an unsigned little endian 32-bit integer that states the length of all
+   ciphertext segments to be sent, in bytes.
+2. If the number of remaining unsent bytes in a message are less than or equal
+   to 65519, encrypt and send them over the network; done.
+3. Otherwise, encrypt and send the first 65519 bytes; return to step 2.
+
+See the subsequent subsections for concrete details.
+
+### 5.3 Message encoding
+This subsection defines pseudocode function `WriteMsg(plaintext)` that takes
+the full Cable Wire Protocol message payload, `plaintext`, as bytes, and writes
+them to the network, performing encryption and fragmentation:
+
+```js
+function WriteMsg (plaintext) {
+  let written = 0
+  while (written < plaintext.length) {
+    let segmentLen = min(65519, plaintext.length - written)
+    let bytes = plaintext.slice(written, segmentLen)
+    let ciphertext = EncryptWithAd(ZERO, bytes)
+    WriteBytes(ciphertext)
+    written += bytes.length
+  }
+}
+```
+
+When a Cable Wire Protocol message, `plaintext` is to be sent, it MUST follow
+these steps:
+
+1. Compute the total length of all of the cipertexts fragments, in bytes, as
+   `len`, a 32-bit unsigned little endian integer.
+
+2. `WriteBytes(len)`
+
+3. `WriteMsg(plaintext)`
+
+If, for example, a `plaintext` of length 90200 were to be encoded & written,
+the first 65519 bytes would first be encrypted and written, followed by the
+remaining 24681 bytes being encrypted and written. Since each ciphertext has an
+extra 16 bytes added of authentication data, and there are two segments
+written, the total written length would be `65519 + 24681 + 16 * 2 = 90232`
+bytes.
+
```

Comment:
I'm glad you named this! I actually originally wrote it to have message lengths encrypted, but figured it would be easy for an attacker to figure out the lengths anyways, so it wasn't worth hiding, since the authentication on the 4-byte length prefix would be an extra 16 bytes. Can you say more about why it makes you sad?

What I realize now is that encrypting the length prefixes here makes the prefix *authenticated*: that lets us preserve data integrity, and also protects against the attacker generating a fake [end-of-stream marker](https://github.com/cabal-club/cable/issues/17).

But it sounds like you're thinking primarily about avoiding leaking message lengths? Wouldn't 256-byte fragments still give a listener a pretty good idea of the lengths of the messages being sent? I think I'm trying to understand what this would protect against. cc @keks on this too.

Reply:
> but figured it would be easy for an attacker to figure out the lengths anyways, so it wasn't worth hiding

Ideally, an attacker would just see a stream of gibberish, with no idea where the message boundaries are. So they'd have no idea whether they are observing a single large message or many small messages. Which is oversimplified, since they can correlate the receiving of some bytes with the subsequent sending of other bytes, but still, this would be a nice property to have.

> Wouldn't 256-byte fragments still give a listener a pretty good idea of the lengths of the messages being sent?

You'd encrypt the length of each fragment, and pad every shorter fragment with zeros. Hence, all the observer gets is a stream of gibberish, with the total number of bytes always being a multiple of 257. But they wouldn't know where in that stream any messages stop and others start.

The important part here is encrypting the lengths, which makes fingerprinting via exact message lengths a lot more difficult. The fixed-size chunking then takes it from "a lot more difficult" to "provably satisfying certain security notions".

You might have a distinctive pattern of messages that could be identified from message lengths alone. For example, a yes-or-no question, followed by a lot of posts that either said "yes" or "no". You could discern the number of yeses and nos from message lengths alone. Or if somebody put the first Harry Potter movie into a single message, then there are probably few other messages that have this exact length. Both of these examples are simplistic and not fully unrealistic, of course, but they sketch why unencrypted lengths make me sad.

Reply:
@AljoschaMeyer @keks This makes a lot of sense -- thank you so much! My understanding is that plaintext lengths could potentially reveal a lot of things:
- that the hosts involved are indeed using cable
- the movement of messages through the network to see who is in the same cabal
- information about the messages themselves by the length (I checked, and, for example, it should be possible to know that someone joined/left a given channel from lengths alone 😹)

Given this, I think it makes sense to encrypt message lengths. I'm going to need to think more about the 256-byte fragment idea.

Reply:
Done! I rewrote the entire section for better clarity.